pydantic/_internal/_core_utils.py

```diff
@@ -416,92 +435,122 @@ def walk_core_schema(schema: core_schema.CoreSchema, f: Walk) -> core_schema.Cor
     return f(schema, _dispatch)
 
 
-def _simplify_schema_references(schema: core_schema.CoreSchema, inline: bool) -> core_schema.CoreSchema:  # noqa: C901
-    all_defs: dict[str, core_schema.CoreSchema] = {}
+class _DefinitionsState(TypedDict):
+    definitions: dict[str, core_schema.CoreSchema]
+    ref_counts: dict[str, int]
+    involved_in_recursion: dict[str, bool]
+    current_recursion_ref_count: dict[str, int]
 
-    def make_result(schema: core_schema.CoreSchema, defs: Iterable[core_schema.CoreSchema]) -> core_schema.CoreSchema:
-        definitions = list(defs)
-        if definitions:
-            return core_schema.definitions_schema(schema=schema, definitions=definitions)
-        return schema
+
+def simplify_schema_references(schema: core_schema.CoreSchema) -> core_schema.CoreSchema:  # noqa: C901
+    """Simplify schema references by:
+    1. Inlining any definitions that are only referenced in one place and are not involved in a cycle.
+    2. Removing any unused `ref` references from schemas.
+    """
+    state = _DefinitionsState(
+        definitions={},
+        ref_counts=defaultdict(int),
+        involved_in_recursion={},
+        current_recursion_ref_count=defaultdict(int),
+    )
```

Comment:
What's the motivation for moving into a state? I guess it makes closures smaller, but causes lots of dictionary lookups?

Reply:
So that it can be easily stored in a single metadata key